### ISSUE TITLE
Running multiple Jenkins instances, i.e. multiple port mapping

### ISSUE THREAD

**Issue by jonbrohauge** [vfarcic/docker-flow-proxy#381]
**Description**
We are trying to run multiple instances of Jenkins while using docker-flow-proxy.
Quick setup overview:
 - Docker Swarm having three managers and multiple workers
 - Jenkins instances running on workers
 - Jenkins agents run on separate AWS ECS cluster

We are trying to configure docker-flow-proxy in our Jenkins service definition, having both ports (8080, 50000) accessible.

**Describe the results you received:**
If creating a Jenkins service with this command:
```
docker service create --name=sample-ci --network=proxy --constraint 'node.role==worker' --label com.df.notify=true --label com.df.distribute=true --label com.df.serviceDomain=sample.ci.mycompany.local --label com.df.port.1=8080 --label com.df.servicePath.1=/ --label com.df.srcPort.2=50000 --label com.df.port.2=50000 --label com.df.servicePath.2=/ --env "GIT_SEED=sample-applications" --detach=true private-docker-repository:5000/mycompany/jenkins-ci:5.0.0-SNAPSHOT
```

The output of the jenkins-agent:
`INFO: Connecting to sample.ci.mycompany.local:50000
Not use http_proxy environment variable which is invalid: no protocol:
Nov 22, 2017 9:48:31 AM hudson.remoting.jnlp.Main$CuiListener error
SEVERE: The server rejected the connection: None of the protocols were accepted
java.lang.Exception: The server rejected the connection: None of the protocols were accepted
        at hudson.remoting.Engine.onConnectionRejected(Engine.java:380)
        at hudson.remoting.Engine.run(Engine.java:352)
`

**Describe the results you expected:**
My artifact get built, i.e. a satisfactory result.

**Additional information you deem important (e.g. issue happens only occasionally):**
Following your Jenkins service yaml from your presentation (vfarcic.github.io/jenkins-swarm/), we can get one instance running, however we need to have more than one instance running on the same swarm. When only having deployed one (1) Jenkins instance without the `--label com.df.XXX.2=` AND not having `BIND_PORTS=50000` on the _docker-flow-proxy_ configuration, thus exposing port 50000 directly, it works.

Any advice on how to get this working will be greatly appreciated.

**Comment by vfarcic** [comment]
Can you please clarify what you mean by multiple Jenkins instances. Are they different Jenkins services or a single service with multiple replicas?

**Comment by jonbrohauge** [comment]
To clarify:
We wish to run multiple instances of Jenkins, i.e. different Jenkins servers, potentially doing different things, and not a High-Availability solution where one Jenkins instance is spread across the swarm.
For example:
Jenkins Instance 1: should be reachable by http://jenkins-1.ci.mycompany.local/
Jenkins Instance 2: should be reachable by http://jenkins-2.ci.mycompany.local/

**Comment by vfarcic** [comment]
That should work :). Can you post your stack and commands so that I try to reproduce the issue?

**Comment by jonbrohauge** [comment]
proxy network is created by: `docker network create --driver overlay --opt encrypted=true proxy`

We deploy our proxy stack: `docker stack deploy -c proxy-stack.yaml proxy`

Our proxy-stack.yaml:
```
version: "3.4"

services:
  proxy:
    image: vfarcic/docker-flow-proxy
    ports:
      - "80:80"
      - "443:443"
      - "50000:50000"
    networks:
      - proxy
    deploy:
      replicas: 3
      placement:
        constraints:
          - node.role == manager
      restart_policy:
        condition: on-failure
        delay: 5s
        max_attempts: 2
    environment:
      LISTENER_ADDRESS: "swarm-listener"
      SERVICE_IGNORE: "true"
      MODE: "swarm"
      BIND_PORTS: 50000

  swarm-listener:
    image: artifactory:5000/vfarcic/docker-flow-swarm-listener
    networks:
      - proxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      DF_NOTIFY_CREATE_SERVICE_URL: "http://proxy:8080/v1/docker-flow-proxy/reconfigure"
      DF_NOTIFY_REMOVE_SERVICE_URL: "http://proxy:8080/v1/docker-flow-proxy/remove"
      SERVICE_IGNORE: "true"
    deploy:
      replicas: 3
      placement:
        constraints:
          - node.role == manager
      restart_policy:
        condition: on-failure
        delay: 5s
        max_attempts: 2

networks:
  proxy:
    external: true
```

**Comment by jonbrohauge** [comment]
Deploying our jenkins instances is as described previously, with the exception of `--name` and `--label com.df.serviceDomain` are customized.

**Comment by matt328** [comment]
I'm having the same issue, it seems if I remap any of the images' ports in the service definition, the proxy service returns 503s.

For example, the jenkins image listens on 8080, and I'd like to run more than one service whose image listens on 8080, so I'd like to do:

```
ports:
  - 8081:8080
```
but any remapping of the ports at the service level causes 503s.

**Comment by vfarcic** [comment]
I haven't yet had time to try the same stack/commands. I'm not sure I'll find a free moment before weekend.

Generally, I'm not sure there's a good reason to publish agent ports through DFP. It is equally simple to simply publish those ports (e.g., 50000) on the service. If agents communicate through the proxy, that communication needs to be TCP since agents do not use HTTP but JNLP. From what I've seen, you're trying to set them up through HTTP. Also, you'd need to make sure that the ports are fixed since Jenkins tends to assign a random number for agents communication.

**Comment by jonbrohauge** [comment]
I missed the part of working with the wrong protocol. It makes sense, since I'm trying to route TCP-traffic through an HTTP-proxy. I think I saw that it is possible to route TCP instead of HTTP in docker-flow-proxy.
One of the good things about docker-flow-proxy is that it takes care of maintaining the Load Balancer Listeners in AWS for me, and thus it would be great to be possible to route through a specific port (50000), so we do not have to open up for almost all ports in the listener.
We have not enabled _random port_ in agent communication. We have it set to port 50000, which is default.

**Comment by jonbrohauge** [comment]
Using  `--label com.df.reqMode.2=tcp` does not work 😞 
These two lines show up in the CloudWatch:
`[WARNING] 326/082127 (28208) : config : 'option forwardfor' ignored for frontend 'tcpFE_50000' as it requires HTTP mode.`
`[WARNING] 326/082127 (28208) : config : 'option forwardfor' ignored for backend 'sample-ci-be50000_2' as it requires HTTP mode.`

So you cannot mix and match protocols...

**Comment by vfarcic** [comment]
I'd suggest to publish agent ports on the service instead. Myself, I use DFP for cases when requests should all go to the same ports (e.g., 80 and 443) and mostly for HTTP. TCP protocol is problematic since it does not have headers so a proxy cannot distinguish one request from another and decide whether to forward them. For example, I do forward HTTP(S) through DFP (https://github.com/vfarcic/docker-flow-hub/blob/master/stack.yml#L5) but keep the port 50000 published directly (https://github.com/vfarcic/docker-flow-hub/blob/master/stack.yml#L8).

P.S. That's my small testing cluster.

**Comment by jonbrohauge** [comment]
What we ended up with is using DFP for the http access to our Jenkins servers. An as you suggested exposing the agent port (50000) on the service directly. We have made a web-based application where one can request a Jenkins instance and thus control the port-mapping so we can ensure that the proper port is set in the Jenkins configuration at boot time, using groovy and rest.
Thanks for your views and giving us some insight into the does and don'ts of DFP.
